modules/core/04-channel/keeper/upgrade.go

```diff
@@ -22,115 +14,64 @@ func (k Keeper) ChanUpgradeInit(
 // An event is emitted for the handshake step.
-func (k Keeper) WriteUpgradeInitChannel(
-	ctx sdk.Context,
-	portID,
-	channelID string,
-	upgradeSequence uint64,
-	channelUpgrade types.Channel,
-) {
+func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID string, currentChannel types.Channel, upgrade types.Upgrade) {
 	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-init")
 
-	k.SetChannel(ctx, portID, channelID, channelUpgrade)
-	k.Logger(ctx).Info("channel state updated", "port-id", portID, "channel-id", channelID, "previous-state", types.OPEN.String(), "new-state", types.INITUPGRADE.String())
-
-	emitChannelUpgradeInitEvent(ctx, portID, channelID, upgradeSequence, channelUpgrade)
-}
-
-// RestoreChannel restores the given channel to the state prior to upgrade.
-func (k Keeper) RestoreChannel(ctx sdk.Context, portID, channelID string, upgradeSequence uint64, err error) error {
-	errorReceipt := types.NewErrorReceipt(upgradeSequence, err)
-	k.SetUpgradeErrorReceipt(ctx, portID, channelID, errorReceipt)
+	currentChannel.State = types.INITUPGRADE
 
-	channel, found := k.GetUpgradeRestoreChannel(ctx, portID, channelID)
-	if !found {
-		return errorsmod.Wrapf(types.ErrChannelNotFound, "channel-id: %s", channelID)
-	}
+	k.SetChannel(ctx, portID, channelID, currentChannel)
+	k.SetUpgrade(ctx, portID, channelID, upgrade)
 
-	k.SetChannel(ctx, portID, channelID, channel)
-	k.DeleteUpgradeRestoreChannel(ctx, portID, channelID)
-	k.DeleteUpgradeTimeout(ctx, portID, channelID)
-
-	module, _, err := k.LookupModuleByChannel(ctx, portID, channelID)
-	if err != nil {
-		return errorsmod.Wrap(err, "could not retrieve module from port-id")
-	}
+	k.Logger(ctx).Info("channel state updated", "port-id", portID, "channel-id", channelID, "previous-state", types.OPEN.String(), "new-state", types.INITUPGRADE.String())
 
-	portKeeper, ok := k.portKeeper.(*portkeeper.Keeper)
-	if !ok {
-		panic("todo: handle this situation")
-	}
+	emitChannelUpgradeInitEvent(ctx, portID, channelID, currentChannel, upgrade)
+}
 
-	cbs, found := portKeeper.Router.GetRoute(module)
+// constructProposedUpgrade returns the proposed upgrade from the provided arguments.
+func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID string, fields types.UpgradeFields, timeout types.UpgradeTimeout) (types.Upgrade, error) {
+	seq, found := k.GetNextSequenceSend(ctx, portID, channelID)
 	if !found {
-		return errorsmod.Wrapf(porttypes.ErrInvalidRoute, "route not found to module: %s", module)
+		return types.Upgrade{}, types.ErrSequenceSendNotFound
 	}
-
-	return cbs.OnChanUpgradeRestore(ctx, portID, channelID)
+	return types.Upgrade{
+		Fields: types.UpgradeFields{
+			Ordering:       fields.Ordering,
+			ConnectionHops: fields.ConnectionHops,
+			Version:        fields.Version,
+		},
+		Timeout: types.UpgradeTimeout{
+			Height:    timeout.Height,
+			Timestamp: timeout.Timestamp,
+		},
+		LatestSequenceSend: seq - 1,
```

Comment:
should we add a comment for why we do the minus 1?

Reply:
We can add a comment if you think it will help, I think the combination of the names `GetNextSequenceSend` and `LatestSequenceSend` make it pretty clear, but I can add a comment to make it extra explicit if you think it makes it more clear.

Reply:
we could also consider adding a new keeper method to sweep it under the rug a little. E.g. `k.GetLastSequenceSend(ctx, portID, channelID)` which would read the same key/value as `GetNextSequenceSend` but handle the off by one.

I think if we do that it can be done later in a different PR.